[Data] Add exception handling for invalid URIs in download operation

xyuzh opened the pull request:
## Description

This PR hardens the download operation planner against bad URIs in two places:

1. **Download phase:** `load_uri_bytes` now gracefully handles invalid/inaccessible URIs and returns `None` instead of crashing the pipeline.
2. **Partition size estimation:** `_estimate_nrows_per_partition` now avoids a `ZeroDivisionError` when *all* sampled URIs fail, and falls back to using the number of rows in the block as the partition size.

Together, these changes make the data pipeline more resilient when dealing with flaky or invalid URLs.

## Changes made

* Wrap `fs.open_input_file(uri_path)` in a `try`/`except` block inside `load_uri_bytes`.

  * On failure, return `None` for that URI.
  * Log a warning with the URI, column name, and error message.
* Update `_estimate_nrows_per_partition` in `PartitionActor` to handle the case where all sampled files fail to return sizes:

  * Detect when no valid sizes are available (e.g., total sampled bytes is zero).
  * Log a warning and fall back to using the number of rows in the block as the partition size instead of dividing by zero.
* Keep existing behavior unchanged when downloads and size sampling succeed.

## Related issues

Closes #58462

## Behavior

### 1. Download errors in `load_uri_bytes`

When a URI download fails (invalid URI, 404, network error, permission error, etc.):

1. The exception is caught inside `load_uri_bytes`.
2. A warning is logged with:

   * The failing URI
   * The source column name
   * The exception message
3. `None` is returned for that specific row’s bytes column.
4. Processing continues for all remaining URIs.

This matches the error-handling pattern used elsewhere in the codebase (e.g., `_sample_sizes`) and avoids failing the entire pipeline due to a few bad URIs.

### 2. Division by zero in `_estimate_nrows_per_partition`

Previously, if **all** sampled URIs failed to produce file sizes, `_estimate_nrows_per_partition` could hit a `ZeroDivisionError` when computing the average bytes per row.

After this change:

1. If every sampled URI fails (no valid sample sizes), the method:

   * Logs a warning indicating that all sampled files failed and sizes could not be estimated.
   * Falls back to using the number of rows in the block as the partition size.
2. No `ZeroDivisionError` is raised; the pipeline continues with a reasonable fallback estimate.

## Testing

### 1. Download error handling

**Manual test script (`test_download_error.py`):**

```python
#!/usr/bin/env python3
"""Test that the try/except block in plan_download_op.py handles download errors correctly."""

import pyarrow as pa
from ray.data.context import DataContext
from ray.data._internal.planner.plan_download_op import download_bytes_threaded

# Test URLs: one valid, one 404
urls = [
    "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTc9APxkj0xClmrU3PpMZglHQkx446nQPG6lA&s",
    "https://www.labelsrus.com/images/products/thumbs/lip-97.jpg",  # 404
]

print(f"\nTesting download with {len(urls)} URLs (1 valid, 1 invalid)...\n")

table = pa.table({"url": urls})
ctx = DataContext.get_current()
results = list(download_bytes_threaded(table, ["url"], ["bytes"], ctx))

result_table = results[0]
for i in range(result_table.num_rows):
    url = result_table["url"][i].as_py()
    bytes_data = result_table["bytes"][i].as_py()

    if bytes_data is None:
        print(f"Row {i}: FAILED (None) - try/except worked ✓")
    else:
        print(f"Row {i}: SUCCESS ({len(bytes_data)} bytes)")
    print(f"  URL: {url[:60]}...")

print("\n✅ Test passed: Failed downloads return None instead of crashing.")
```

**Before this change**

* The worker logs a warning (`Caught exception in transforming worker!`).
* `download_bytes_threaded` raises a `FileNotFoundError` from `fs.open_input_file(...)`.
* The entire pipeline and script fail with a traceback.

**After this change**

* A warning is logged from `plan_download_op.py` indicating the failed URI and column.
* The valid URL row succeeds with non-empty bytes.
* The invalid URL row has `bytes == None` and is reported as `FAILED (None) - try/except worked ✓`.
* The script completes successfully:

```text
Row 0: SUCCESS (...)
Row 1: FAILED (None) - try/except worked ✓
✅ Test passed: Failed downloads return None instead of crashing.
```

### 2. ZeroDivisionError fix in `_estimate_nrows_per_partition`

**Manual test script (`test_divide_zero.py`):**

```python
#!/usr/bin/env python3
"""Test division by zero fix when ALL sampled URLs fail."""

import pyarrow as pa
from ray.data.context import DataContext
from ray.data._internal.planner.plan_download_op import PartitionActor

# Test URLs: ALL invalid (404) - this triggers the bug
# The bug only happens when ALL sampled files fail to return sizes
urls = [
    "https://www.labelsrus.com/images/products/thumbs/lip-97.jpg",  # 404
    "https://example.com/nonexistent/image.jpg",  # 404
]

print(f"\nTesting partition size estimation with {len(urls)} invalid URLs...\n")

try:
    table = pa.table({"url": urls})
    ctx = DataContext.get_current()
    partition_actor = PartitionActor(uri_column_names=["url"], data_context=ctx)

    estimated_rows = partition_actor._estimate_nrows_per_partition(table)
    print(f"✅ Success! Estimated rows per partition: {estimated_rows}")
    print("   The fix prevents division by zero when all files fail.\n")

except ZeroDivisionError as e:
    print(f"❌ FAILED: ZeroDivisionError - {e}")
except Exception as e:
    print(f"❌ FAILED: {type(e).__name__} - {e}\n")
```

**Before this change**

```text
Testing partition size estimation with 2 invalid URLs...

❌ FAILED: ZeroDivisionError - division by zero
```

**After this change**

```text
Testing partition size estimation with 2 invalid URLs...

2025-11-08 20:39:42,725 WARNING plan_download_op.py:290 -- Unable to estimate file sizes for URIs (all sampled files failed). Falling back to using the number of rows in the block as the partition size.
✅ Success! Estimated rows per partition: 2
   The fix prevents division by zero when all files fail.
```

This confirms that:

* The `ZeroDivisionError` is eliminated when all sampled files fail.
* A clear warning is logged about the fallback behavior.
* The method returns a sane partition-size estimate based on the row count, and the pipeline keeps running.

xyuzh committed:
Add exception handling for URI download failures

- Added try-except block in load_uri_bytes function to handle invalid URIs
- Catch specific exceptions (OSError, pa.ArrowException) for better error handling
- Failed downloads now return None instead of crashing the pipeline
- Added warning log message for failed URI downloads

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
Fix divide by zero error when all URI size estimations fail

- Modified _sample_sizes to append 0 for failed URLs instead of discarding them
- Failed URLs (returning None or raising exceptions) now contribute 0 to size estimates
- Ensures file_sizes list length matches number of sampled URIs
- Allows avg_nbytes_per_row == 0 check to catch the all-failures case
- Maintains empty row_sizes check for edge case when no URIs are sampled

Signed-off-by: xyuzh <xinyzng@gmail.com>

robertnishihara commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -189,8 +189,14 @@ def download_bytes_threaded(
+                try:
+                    with fs.open_input_file(uri_path) as f:
+                        yield f.read()
+                except (OSError, pa.ArrowException) as e:
```
I think we should just narrowly catch the `FileNotFoundError` to avoid masking other errors.

What's the pyarrow exception for? Can we get rid of that?

xyuzh replied:
The OSError covers:
`FileNotFoundError`
`PermissionError`
`IsADirectoryError`
`NotADirectoryError`
`BlockingIOError`
`TimeoutError`
Those are general error we all want to handle for downloading, I would remove pa.ArrowException which is unnecessary

robertnishihara commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -189,8 +189,14 @@ def download_bytes_threaded(
+                    with fs.open_input_file(uri_path) as f:
+                        yield f.read()
+                except (OSError, pa.ArrowException) as e:
+                    logger.warning(
```
This seems too verbose to me. A dataset could have 100s of millions of URLs that fail.

xyuzh replied:
Agree, I have removed the logging here

robertnishihara commented:
Not for this PR, @richardliaw looking at the code, it seems to me that the logic when there are more than one uri columns is incorrect. I could be wrong, but it seems like this zip call will fail
https://github.com/ray-project/ray/blob/654feda85f11b77aaa07a1c96eb11cc83195e161/python/ray/data/_internal/planner/plan_download_op.py#L277
the lists will have different lengths because in `_sample_sizes`, we discard errors
https://github.com/ray-project/ray/blob/654feda85f11b77aaa07a1c96eb11cc83195e161/python/ray/data/_internal/planner/plan_download_op.py#L325-L326
Why do we support multiple uri columns? Can we just get rid of all that logic (it seems like we don't use it and it looks buggy)?

xyuzh committed:
Add comprehensive edge case tests for URI download failures

- Add test_download_expression_with_invalid_uris: Tests exception handling
  for URIs that fail to download (OSError), verifying None is returned
  instead of crashing (covers commit c9d91080fb)

- Add test_download_expression_all_size_estimations_fail: Tests divide-by-zero
  fix when all URI size estimations fail during partitioning, ensuring fallback
  to row count works correctly (covers commit 095973428f)

- Add test_download_expression_mixed_valid_and_invalid_size_estimation: Tests
  partial failures during size estimation with mix of valid/invalid URIs

These tests properly cover the code paths that were previously untested:
- load_uri_bytes exception handling (lines 192-196)
- avg_nbytes_per_row == 0 check (lines 285-291)
- Failed size estimation handling (lines 326-332)

Signed-off-by: xyuzh <xinyzng@gmail.com>

robertnishihara commented:
The tests look good, thanks for adding those.

robertnishihara commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -322,9 +326,9 @@ def get_file_size(uri_path, fs):
+                    file_sizes.append(size if size is not None else 0)
                 except Exception as e:
                     logger.warning(f"Error fetching file size for download: {e}")
+                    file_sizes.append(0)
```
This is a really interesting catch. However, I think we should solve this separately by removing support for multiple columns.

richardliaw commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -189,8 +189,11 @@ def download_bytes_threaded(
+                    with fs.open_input_file(uri_path) as f:
+                        yield f.read()
+                except OSError as _:
+                    yield None
```
can you except and logging.debug the error / file path

this won't show by default but will be useful for debugging

richardliaw commented:
> I could be wrong, but it seems like this zip call will fail

the zip call won't fail but it won't be accurate. it will just truncate when one finishes.

zip([1,2,3], [1,2]) -> [(1,1), (2,2)]

xyuzh commented on python/ray/data/tests/test_download_expression.py:
```diff
@@ -294,6 +294,115 @@ def test_download_expression_with_null_uris(self):
+    def test_download_expression_with_invalid_uris(self, tmp_path):
+        """Test download expression with URIs that fail to download.
+
+        This tests the exception handling in load_uri_bytes (commit c9d91080fb)
```
removed

xyuzh committed:
Fix trailing whitespace in test docstrings and comments

Signed-off-by: xyuzh <xinyzng@gmail.com>

bveeramani commented on python/ray/data/tests/test_download_expression.py:
```diff
@@ -294,6 +294,117 @@ def test_download_expression_with_null_uris(self):
+
+        # Second and third URIs should fail gracefully (return None)
+        assert results[1]["bytes"] is None
+        assert results[2]["bytes"] is None
```
Fixed by https://github.com/ray-project/ray/pull/58518
